lib/cache.js

```diff
@@ -0,0 +1,322 @@
+    refreshRate: refreshRate,
+    hostname: skimdb.hostname,
+    protocol: skimdb.protocol,
+    port: skimdb.port
+  });
+
+  cache.createKeyStream = function(prefix) {
+    // concat-stream -- concat-strean
+    const start = prefix;
+    const end = start.slice(0, prefix.length - 1)
+              + String.fromCharCode(prefix.charCodeAt(prefix.length - 1) + 1);
+
+    return this.db.createKeyStream({ start: start, end: end });
+  };
+
+  cache._destroy = cache.destroy;
+
+  cache.destroy = function() {
+    this._pubStream.destroy();
+    return cache._destroy();
+  };
+
+  cache._cullHandler = function(data) {
+    // I just happen to know that the name (should, anyway) already has the
+    // format of the key prior to the '@'.
+    const prefix = data.doc.name;
+
+    const dist_tags = data.doc['dist-tags'];
+
+    // Should invalidate anything that has a dist-tag
+    // TODO: Mad overkill for any project with tons of dist-tags, any way to
+    // figure out what the update was?
+    Object.keys(dist_tags).forEach((tag) => {
+      const module_version = dist_tags[tag];
+
+      const stream = this.createKeyStream(prefix);
+
+      stream.on('data', (key) => {
+        log.info('cull stream on data', key);
+
+        // Grab the semver range used in the key
+        const module_semver = key.split('@').pop();
+
+        // If this alias has been invalidated by the tagged version...
+        if (semver.satisfies(module_version, module_semver) || module_semver == tag) {
+          log.info(`culling \`${key}\`...`);
+          this._del(key);
+        }
+      });
+    });
+  };
+  pubStream.on('data', (data) => cache._cullHandler(data));
+
+  pubStream.on('error', (err) => {
+    log.info('pubstream error:', err.message);
+    log.info(err.stack);
+  });
+
+  cache._pubStream = pubStream;
+
+  return cache;
+}
+
+class Cache {
+  constructor(name, db, opts) {
+    opts = opts || {}
+    this.db = db.sublevel(name);
+    this.name = name;
+    this.hashfxn = opts.hashfxn || opts.hashFunction || defaultHashFxn
+    this.ttl = opts.ttl;
+  }
+
+  destroy() {
+    return Promise.fromCallback((cb) => this.db.close(cb));
+  }
+
+  _get(key) {
+    return Promise.fromCallback((cb) => this.db.get(key, cb))
+      .then((v) => {
+        try {
+          return JSON.parse(v)
+        }
+        catch (parseError) {
+          parseError.name = 'JSONParseError';
+          throw parseError;
+        }
+      });
+  }
+
+  get(key) {
+    return Promise.try(() => {
+      return this._get(this.hashfxn(key));
+    });
+  }
+
+  _put(key, val, opts) {
+    return Promise.try(() => {
+      return Promise.fromCallback((cb) => this.db.put(key, JSON.stringify(val), opts, cb));
+    });
+  }
+
+  put(key, val, opts) {
+    return Promise.try(() => {
+      return this._put(this.hashfxn(key), val, opts);
+    });
+  }
+
+  _del(key) {
+    return Promise.try(() => {
+      return Promise.fromCallback((cb) => this.db.del(key, cb));
```

Comment:
why do you need to wrap this in Promise.try? `Promise.fromCallback` surely does the promise wrapping for you